### PR TITLE
Draft: Add Machine Spec to Core

### PR DESCRIPTION
**Pull request by jonathan-innis**
Fixes # 

**Description**

**How was this change tested?**

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/apis/v1alpha1/machine.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	// purposes in that pods are not required to tolerate a StartupTaint in order to have nodes provisioned for them.
+	// +optional
+	StartupTaints []v1.Taint `json:"startupTaints,omitempty"`
+	// Requirements are layered with Labels and applied to every node.
+	Requirements []v1.NodeSelectorRequirement `json:"requirements,omitempty"`
+	// Kubelet are options passed to the kubelet when provisioning nodes
+	// +optional
+	Kubelet *v1alpha5.KubeletConfiguration `json:"kubelet,omitempty"`
+	// NodeTemplateRef is a reference to an object that defines provider specific configuration
+	NodeTemplateRef *v1alpha5.ProviderRef `json:"nodeTemplateRef,omitempty"`
```
Comment (ellistarn) [inline pkg/apis/v1alpha1/machine.go:43]: I'm now wondering if NodeTemplate is even the right name. AWSMachineTemplate?
https://cluster-api-aws.sigs.k8s.io/topics/images/custom-amis.html
Odear.

```suggestion
	MachineTemplateRef *v1alpha1.ObjectReference `json:"nodeTemplateRef,omitempty"`
```

File: pkg/apis/v1alpha1/machine.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	// Provider contains fields specific to your cloudprovider.
+	// Deprecated: NodeTemplateRef is favored over Provider
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Provider *v1alpha5.Provider `json:"provider,omitempty"`
```
Comment (ellistarn) [inline pkg/apis/v1alpha1/machine.go:47]: Do we need to support this?

File: pkg/apis/v1alpha1/machine.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+// MachineList contains a list of Provisioner
+// +kubebuilder:object:root=true
+type MachineList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Machine `json:"items"`
+}
+
+func (in *Machine) SchedulingRequirements() scheduling.Requirements {
```
Comment (ellistarn) [inline pkg/apis/v1alpha1/machine.go:70]:
```suggestion
func (in *Machine) Requirements() scheduling.Requirements {
```

File: pkg/apis/v1alpha1/machine.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	return scheduling.NewNodeSelectorRequirements(in.Spec.Requirements...)
+}
+
+func (in *Machine) ProvisionerOwner() (client.ObjectKey, bool) {
```
Comment (ellistarn) [inline pkg/apis/v1alpha1/machine.go:74]:
```suggestion
func (in *Machine) Owner(kind string) (client.ObjectKey, bool) {
```

File: pkg/apis/v1alpha1/machine_status.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+package v1alpha1
+
+import (
+	"github.com/samber/lo"
+	"knative.dev/pkg/apis"
+)
+
+// MachineStatus defines the observed state of Machine
+type MachineStatus struct {
+	NodeName string `json:"nodeName,omitempty"`
```
Comment (ellistarn) [inline pkg/apis/v1alpha1/machine_status.go:24]: I'm not sure we need this one.

File: pkg/apis/v1alpha1/machine_status.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+	// +optional
+	Conditions apis.Conditions `json:"conditions,omitempty"`
+}
+
+func (in *Machine) StatusConditions() apis.ConditionManager {
+	return apis.NewLivingConditionSet(
+		MachineCreated,
+		MachineRegistered,
+		MachineHealthy,
+		MachineInitialized,
```
Comment (ellistarn) [inline pkg/apis/v1alpha1/machine_status.go:37]:
```suggestion
		MachineInitialized,
		MachineHealthy,
```

File: pkg/apis/v1alpha1/machine_status.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+	ProviderID string `json:"providerID,omitempty"`
+	// Conditions contains signals for health and readiness
+	// +optional
+	Conditions apis.Conditions `json:"conditions,omitempty"`
+}
+
+func (in *Machine) StatusConditions() apis.ConditionManager {
+	return apis.NewLivingConditionSet(
+		MachineCreated,
+		MachineRegistered,
```
Comment (ellistarn) [inline pkg/apis/v1alpha1/machine_status.go:35]:
```suggestion
		NodeRegistered,
```

File: pkg/apis/v1alpha1/machine_status.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	MachineRegistered  apis.ConditionType
+	MachineHealthy     apis.ConditionType
+	MachineInitialized apis.ConditionType
+)
+
+func (in *Machine) GetConditions() apis.Conditions {
+	return in.Status.Conditions
+}
+
+func (in *Machine) GetCondition(cond apis.ConditionType) (apis.Condition, bool) {
```
Comment (ellistarn) [inline pkg/apis/v1alpha1/machine_status.go:52]: Do we need this one 🤔

File: pkg/cloudprovider/fake/cloudprovider.go [hunk 9]
```diff
@@ -52,17 +54,26 @@ func NewCloudProvider() *CloudProvider {
...
+	c.CreateCalls = append(c.CreateCalls, machine)
 	if len(c.CreateCalls) > c.AllowedCreateCalls {
 		c.mu.Unlock()
 		return &v1.Node{}, fmt.Errorf("erroring as number of AllowedCreateCalls has been exceeded")
 	}
 	c.mu.Unlock()
 
-	name := test.RandomName()
-	instanceType := nodeRequest.InstanceTypeOptions[0]
+	requirements := scheduling.NewNodeSelectorRequirements(machine.Spec.Requirements...)
```
Comment (ellistarn) [inline pkg/cloudprovider/fake/cloudprovider.go:66]: Do you need this if you have machine.Requirements()? Probably safer to not have v1alpha1 rely on the scheduling package.

File: pkg/apis/v1alpha1/machine.go (hunk 2 above)
Comment (jonathan-innis) [reply pkg/apis/v1alpha1/machine.go:47]: So we're just saying that with this change, we will just start breaking customers that are relying on Provider if they haven't migrated?

**Pull request closed without merging**